File: Services/pkg/models/permission.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Permission struct {
	ID        uint      `json:"id" gorm:"primaryKey;autoIncrement;index"`
	Group     string    `json:"group" gorm:"column:pkg;not null;index"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}
type Role struct {
	ID        uint      `json:"id" gorm:"primaryKey;autoIncrement;index"`
	Name      string    `json:"name" gorm:"unique;not null"`
	CreatedAt time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}
type RolePermission struct {
	ID           uint       `json:"id" gorm:"primaryKey;autoIncrement;index"`
	RoleID       uint       `json:"role_id" gorm:"not null;index"`
	PermissionID uint       `json:"permission_id" gorm:"not null;index"`
	CreatedAt    time.Time  `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt    time.Time  `json:"updated_at" gorm:"autoUpdateTime"`
	Role         Role       `json:"-" gorm:"foreignKey:RoleID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	Permission   Permission `json:"-" gorm:"foreignKey:PermissionID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
}
type UserRole struct {
	ID        uint      `json:"id" gorm:"primaryKey;autoIncrement;index"`
	UserID    uint      `json:"user_id" gorm:"not null;index"`
	RoleID    uint      `json:"role_id" gorm:"not null;index"`
	CreatedAt time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`
	Role      Role      `json:"-" gorm:"foreignKey:RoleID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	User      User      `json:"-" gorm:"foreignKey:UserID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
}

func MigrateTablePermissions(db *gorm.DB) error {
	err := db.AutoMigrate(&Permission{}, &Role{}, &RolePermission{}, &UserRole{})
	if err != nil {
		return err
	}
	return err
}

type PermissionGroup_ string
type PermissionName_ string

const (
	Create PermissionName_ = "create"
	Read   PermissionName_ = "read"
	Update PermissionName_ = "update"
	Delete PermissionName_ = "delete"
	List   PermissionName_ = "list"
	Me     PermissionName_ = "me"
)
const (
	UserGroup           PermissionGroup_ = "user"
	RoleGroup           PermissionGroup_ = "role"
	PermissionGroup     PermissionGroup_ = "permission"
	RolePermissionGroup PermissionGroup_ = "role_permission"
	UserRoleGroup       PermissionGroup_ = "user_role"
)

func PermissionGroupName(group PermissionGroup_, name PermissionName_) string {
	return string(group) + ":" + string(name)
}

File: hono-clean-architecture/pkg/models/permission.ts
import { pgTable, serial, varchar, timestamp, integer } from "drizzle-orm/pg-core";
import { InferSelectModel, InferInsertModel } from "drizzle-orm";
import { users } from "./user";

export const permissions = pgTable("permissions", {
  id: serial("id").primaryKey(),
  group: varchar("pkg", { length: 255 }).notNull(),
  name: varchar("name", { length: 255 }).notNull(),
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at").defaultNow().notNull(),
});

export type Permission = InferSelectModel<typeof permissions>;
export type NewPermission = InferInsertModel<typeof permissions>;

export const roles = pgTable("roles", {
  id: serial("id").primaryKey(),
  name: varchar("name", { length: 255 }).notNull().unique(),
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at").defaultNow().notNull(),
});

export type Role = InferSelectModel<typeof roles>;
export type NewRole = InferInsertModel<typeof roles>;

export const rolePermissions = pgTable("role_permissions", {
  id: serial("id").primaryKey(),
  roleId: integer("role_id")
    .notNull()
    .references(() => roles.id, { onDelete: "cascade", onUpdate: "cascade" }),
  permissionId: integer("permission_id")
    .notNull()
    .references(() => permissions.id, {
      onDelete: "cascade",
      onUpdate: "cascade",
    }),
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at").defaultNow().notNull(),
});

export type RolePermission = InferSelectModel<typeof rolePermissions>;
export type NewRolePermission = InferInsertModel<typeof rolePermissions>;

export const userRoles = pgTable("user_roles", {
  id: serial("id").primaryKey(),
  userId: integer("user_id")
    .notNull()
    .references(() => users.id, { onDelete: "cascade", onUpdate: "cascade" }),
  roleId: integer("role_id")
    .notNull()
    .references(() => roles.id, { onDelete: "cascade", onUpdate: "cascade" }),
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at").defaultNow().notNull(),
});

export type UserRole = InferSelectModel<typeof userRoles>;
export type NewUserRole = InferInsertModel<typeof userRoles>;

export const PermissionName = {
  Create: "create",
  Read: "read",
  Update: "update",
  Delete: "delete",
  List: "list",
  Me: "me",
} as const;

export type PermissionNameType = (typeof PermissionName)[keyof typeof PermissionName];

export const PermissionGroupType = {
  User: "user",
  Role: "role",
  Permission: "permission",
  RolePermission: "role_permission",
  UserRole: "user_role",
} as const;

export type PermissionGroupTypeValue =
  (typeof PermissionGroupType)[keyof typeof PermissionGroupType];

export function permissionGroupName(
  group: PermissionGroupTypeValue,
  name: PermissionNameType
): string {
  return `${group}:${name}`;
}
